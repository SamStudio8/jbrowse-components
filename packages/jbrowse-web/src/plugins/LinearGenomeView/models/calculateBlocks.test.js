import {
  calculateBlocksForward,
  calculateBlocksReversed,
} from './calculateBlocks'

describe('block calculation', () => {
  it('can calculate some blocks 1', () => {
    const blocks = calculateBlocksForward({
      bpPerPx: 1,
      width: 800,
      offsetPx: 0,
      displayedRegions: [
        { assembly: 'volvox', refName: 'ctgA', start: 0, end: 10000 },
      ],
    })
    expect(blocks).toMatchSnapshot()
  })

  it('can calculate some blocks 2', () => {
    const blocks = calculateBlocksForward({
      bpPerPx: 1,
      width: 800,
      offsetPx: 30,
      displayedRegions: [
        { assembly: 'volvox', refName: 'ctgA', start: 0, end: 100 },
        { assembly: 'volvox', refName: 'ctgB', start: 100, end: 200 },
      ],
    })
    expect(blocks).toMatchSnapshot()
  })

  it('can calculate some blocks 3', () => {
    const blocks = calculateBlocksForward({
      bpPerPx: 1,
      width: 800,
      offsetPx: 1000,
      displayedRegions: [
        { assembly: 'volvox', refName: 'ctgA', start: 0, end: 100 },
        { assembly: 'volvox', refName: 'ctgB', start: 100, end: 200 },
      ],
    })
    expect(blocks).toEqual([])
  })

  it('can calculate some blocks 4', () => {
    const blocks = calculateBlocksForward(
      {
        bpPerPx: 1,
        width: 800,
        offsetPx: -1000,
        displayedRegions: [
          { assembly: 'volvox', refName: 'ctgA', start: 0, end: 100 },
          { assembly: 'volvox', refName: 'ctgB', start: 100, end: 200 },
        ],
      },
      {
        testEnv: true,
      },
    )
    expect(blocks).toEqual([])
  })

  it('can calculate some blocks 5', () => {
    const blocks = calculateBlocksForward({
      bpPerPx: 1,
      width: 800,
      offsetPx: 5000,
      displayedRegions: [
        { assembly: 'volvox', refName: 'ctgA', start: 0, end: 10000 },
        { assembly: 'volvox', refName: 'ctgB', start: 100, end: 10000 },
      ],
    })
    expect(blocks).toMatchSnapshot()
  })

  it('can calculate some blocks 6', () => {
    const blocks = calculateBlocksForward({
      bpPerPx: 1,
      width: 800,
      offsetPx: 0,
      displayedRegions: [
        {
          assembly: 'volvox',
          refName: 'ctgA',
          start: 0,
          end: 200,
        },
        { assembly: 'volvox', refName: 'ctgB', start: 0, end: 1000 },
      ],
    })
    expect(blocks).toMatchSnapshot()
  })

  it('can calculate some blocks 7', () => {
    const blocks = calculateBlocksForward({
      bpPerPx: 1,
      width: 800,
      offsetPx: 801,
      displayedRegions: [
        { assembly: 'volvox', refName: 'ctgA', start: 0, end: 200 },
        { assembly: 'volvox', refName: 'ctgB', start: 0, end: 1000 },
      ],
    })
    expect(blocks).toMatchSnapshot()
  })

  it('can calculate some blocks 8', () => {
    const blocks = calculateBlocksForward({
      bpPerPx: 1,
      width: 800,
      offsetPx: 1600,
      displayedRegions: [
        { assembly: 'volvox', refName: 'ctgA', start: 0, end: 200 },
        { assembly: 'volvox', refName: 'ctgB', start: 0, end: 10000000 },
      ],
      configuration: 'fakeReference',
    })
    expect(blocks).toMatchSnapshot()
  })

  it('can calculate some blocks 9', () => {
    const blocks = calculateBlocksForward({
      width: 800,
      offsetPx: 1069,
      bpPerPx: 2,
      displayedRegions: [
        {
          assembly: 'volvox',
          refName: 'ctgA',
          start: 0,
          end: 50000,
        },
        { assembly: 'volvox', refName: 'ctgB', start: 0, end: 300 },
      ],
      configuration: 'fakeReference',
    })
    expect(blocks).toMatchSnapshot()
  })
})

describe('reverse block calculation', () => {
  test('1', () => {
    const blocks = calculateBlocksReversed({
      bpPerPx: 1,
      width: 800,
      offsetPx: 0,
      displayedRegions: [
        { assembly: 'volvox', refName: 'ctgA', start: 0, end: 10000 },
      ],
    })
    expect(blocks).toMatchSnapshot()
  })
})
