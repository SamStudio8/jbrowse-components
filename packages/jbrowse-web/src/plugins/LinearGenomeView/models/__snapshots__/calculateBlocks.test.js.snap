// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`block calculation can calculate some blocks 1 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 800,
    "isLeftEndOfDisplayedRegion": true,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:1-800",
    "offsetPx": 0,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 0,
    "widthPx": 800,
  },
  Object {
    "assembly": "volvox",
    "end": 1600,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:801-1600",
    "offsetPx": 800,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 800,
    "widthPx": 800,
  },
]
`;

exports[`block calculation can calculate some blocks 2 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 100,
    "isLeftEndOfDisplayedRegion": true,
    "isRightEndOfDisplayedRegion": true,
    "key": "volvox:ctgA:1-100",
    "offsetPx": 0,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 100,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 0,
    "widthPx": 100,
  },
  Object {
    "assembly": "volvox",
    "end": 200,
    "isLeftEndOfDisplayedRegion": true,
    "isRightEndOfDisplayedRegion": true,
    "key": "volvox:ctgB:101-200",
    "offsetPx": 200,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 200,
      "refName": "ctgB",
      "start": 100,
    },
    "refName": "ctgB",
    "start": 100,
    "widthPx": 100,
  },
]
`;

exports[`block calculation can calculate some blocks 5 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 5600,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:4801-5600",
    "offsetPx": 4800,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 4800,
    "widthPx": 800,
  },
  Object {
    "assembly": "volvox",
    "end": 6400,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:5601-6400",
    "offsetPx": 5600,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 5600,
    "widthPx": 800,
  },
]
`;

exports[`block calculation can calculate some blocks 6 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 200,
    "isLeftEndOfDisplayedRegion": true,
    "isRightEndOfDisplayedRegion": true,
    "key": "volvox:ctgA:1-200",
    "offsetPx": 0,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 200,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 0,
    "widthPx": 200,
  },
  Object {
    "assembly": "volvox",
    "end": 800,
    "isLeftEndOfDisplayedRegion": true,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgB:1-800",
    "offsetPx": 200,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 1000,
      "refName": "ctgB",
      "start": 0,
    },
    "refName": "ctgB",
    "start": 0,
    "widthPx": 800,
  },
]
`;

exports[`block calculation can calculate some blocks 7 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 800,
    "isLeftEndOfDisplayedRegion": true,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgB:1-800",
    "offsetPx": 200,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 1000,
      "refName": "ctgB",
      "start": 0,
    },
    "refName": "ctgB",
    "start": 0,
    "widthPx": 800,
  },
  Object {
    "assembly": "volvox",
    "end": 1000,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": true,
    "key": "volvox:ctgB:801-1000",
    "offsetPx": 1000,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 1000,
      "refName": "ctgB",
      "start": 0,
    },
    "refName": "ctgB",
    "start": 800,
    "widthPx": 200,
  },
]
`;

exports[`block calculation can calculate some blocks 8 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 1600,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgB:801-1600",
    "offsetPx": 1000,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000000,
      "refName": "ctgB",
      "start": 0,
    },
    "refName": "ctgB",
    "start": 800,
    "widthPx": 800,
  },
  Object {
    "assembly": "volvox",
    "end": 2400,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgB:1601-2400",
    "offsetPx": 1800,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000000,
      "refName": "ctgB",
      "start": 0,
    },
    "refName": "ctgB",
    "start": 1600,
    "widthPx": 800,
  },
]
`;

exports[`block calculation can calculate some blocks 9 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 3200,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:1601-3200",
    "offsetPx": 800,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 50000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 1600,
    "widthPx": 800,
  },
  Object {
    "assembly": "volvox",
    "end": 4800,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:3201-4800",
    "offsetPx": 1600,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 50000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 3200,
    "widthPx": 800,
  },
]
`;

exports[`reverse block calculation 1 1`] = `
Array [
  Object {
    "assembly": "volvox",
    "end": 10000,
    "isLeftEndOfDisplayedRegion": true,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:9201-10000",
    "offsetPx": 0,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 9200,
    "widthPx": 800,
  },
  Object {
    "assembly": "volvox",
    "end": 9200,
    "isLeftEndOfDisplayedRegion": false,
    "isRightEndOfDisplayedRegion": false,
    "key": "volvox:ctgA:8401-9200",
    "offsetPx": 800,
    "parentRegion": Object {
      "assembly": "volvox",
      "end": 10000,
      "refName": "ctgA",
      "start": 0,
    },
    "refName": "ctgA",
    "start": 8400,
    "widthPx": 800,
  },
]
`;
